### PR TITLE
prevent rsync failure.

### LINKED ISSUE #2442: Guessing that files are to be installed by rsync broken in Rose2
At Rose2 some examples of fileinstall source paths behave differently to Rose1:

```
$cat rose-app.conf 
[command]
default=echo "Hello Magrathea!"

[file:file_rsync]
source=exvcylcdev01:file_to_rsync

[file:file_local]
source=/home/h02/tpilling/file_to_get_from_local_fs

[file:file_local_rsync]
source=mycomputer:/home/h02/tpilling/file_to_get_from_local_fs

$rose app-run -C . -v -i
...
[INFO] source: /home/h02/tpilling/file_to_get_from_local_fs
[INFO] install: file_local
[INFO]     source: /home/h02/tpilling/file_to_get_from_local_fs
[INFO] source: vld398:/home/h02/tpilling/file_to_get_from_local_fs
[INFO] install: file_local_rsync
[INFO]     source: vld398:/home/h02/tpilling/file_to_get_from_local_fs
[INFO] source: exvcylcdev01:file_to_rsync
[INFO] install: file_rsync
[INFO]     source: exvcylcdev01:file_to_rsync
[INFO] command: echo "Hello Magrathea!"
```

At Rose2:
```
(cylc8) $rose app-run -C . -v -i
...
[FAIL] file:file_local_rsync=source=mycomputer:/home/h02/tpilling/file_to_get_from_local_fs: bad or missing value
```

### PR DESCRIPTION
**Pull request by wxtim**
Fixes #2442 .

A script written in python2 and unconverted (bc it was a raw script) was being run with python 3. As a result it was returning a tuple when it shouldn't have been.

The script created by metomi/rose/2019.01.x:lib/python/rose/loc_handlers/rsync.py::lines 67-92 should be:
- [x] Stored somewhere in the codebase and read into the stdin at line 94.
- [x] Tested.
- [x] Run using Python3

### CODE REVIEW

File: metomi/rose/loc_handlers/rsync.py [hunk 1]
```diff
@@ -63,34 +63,11 @@ def parse(self, loc, _):
...
-            name = os.path.join(dirpath, filename)
-            stat = os.stat(name)
-            print(oct(stat.st_mode), stat.st_mtime, stat.st_size, name)
-elif os.path.isfile(path):
-    print(str_blob)
-    stat = os.stat(path)
-    print(oct(stat.st_mode), stat.st_mtime, stat.st_size, path)
-""")
+        temp_file.write(
+            (Path(__file__).parent / 'rsync_remote_check.py').read_bytes()
```
Comment (oliver-sanders) [inline metomi/rose/loc_handlers/rsync.py:67]: Can import the path from the installed file directly

```python
from metomi.rose.loc_handlers.rsync_remote_check import __file__ as rsync_remote_check_file
# ...
Path(rsync_remote_check_file).read_bytes()
```

File: metomi/rose/loc_handlers/rsync_remote_check.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Rose. If not, see <http://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+"""
+A script for running over ssh to establish whether a file
+is available for installation.
+
+This is a Python file but we read it and pass it to stdout
```
Comment (oliver-sanders) [inline metomi/rose/loc_handlers/rsync_remote_check.py:24]:
```suggestion
This is a Python file but we read it and pass it to stdin
```

File: metomi/rose/loc_handlers/rsync_remote_check.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                stat = os.stat(name)
+                print((oct(stat.st_mode), stat.st_mtime, stat.st_size, name))
+    elif os.path.isfile(path):
+        print(str_blob)
+        stat = os.stat(path)
+        print(oct(stat.st_mode), stat.st_mtime, stat.st_size, path)
+
+
+if __name__ == '__main__':
+    main()
```
Comment (oliver-sanders) [inline metomi/rose/loc_handlers/rsync_remote_check.py:64]: `\n`

File: metomi/rose/tests/test_loc_handlers_rsync_remote_check.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import sys
+
+from ast import literal_eval
+
+from metomi.rose.loc_handlers.rsync_remote_check import main
+
+
+def test_check_file(monkeypatch, capsys, tmp_path):
+    content = 'blah'
+    permission_level = '0o100777'
```
Comment (oliver-sanders) [inline metomi/rose/tests/test_loc_handlers_rsync_remote_check.py:30]: Not sure what the leading `100` does?

Not keen on creating 777 files! Can we downgrade to 700, 644 or something less scary.

File: metomi/rose/tests/test_loc_handlers_rsync_remote_check.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    dirpath.chmod(int(folder_permission_level, base=8))
+    monkeypatch.setattr(
+        sys, 'argv', ['ignored', str(dirpath), 'blob', 'tree']
+    )
+    main()
+    captured = capsys.readouterr()
+    assert captured.out.splitlines()[0] == 'tree'
+    mode, _, size, path = literal_eval(captured.out.splitlines()[1])
+    assert path == str((dirpath/'more.stuff'))
+    assert mode == '0o100644'
```
Comment (oliver-sanders) [inline metomi/rose/tests/test_loc_handlers_rsync_remote_check.py:63]: I think this 644 is testing how your system is configured (new files aren't necessarily be created with 644 permissions), perhaps explicitly chmod this file.

**Review by oliver-sanders — COMMENTED** [review]
Good catch.

File: metomi/rose/loc_handlers/rsync.py (hunk 1 above)
Comment (wxtim) [reply metomi/rose/loc_handlers/rsync.py:67]: Much tidier. Thanks.

File: metomi/rose/tests/test_loc_handlers_rsync_remote_check.py (hunk 4 above)
Comment (wxtim) [reply metomi/rose/tests/test_loc_handlers_rsync_remote_check.py]: Totally. In one case 644 was fine, in another 700 was as low as I could go without upsetting Pytest's cleanup.

File: metomi/rose/tests/test_loc_handlers_rsync_remote_check.py (hunk 5 above)
Comment (wxtim) [reply metomi/rose/tests/test_loc_handlers_rsync_remote_check.py]: Fixed, I believe.

**Comment by oliver-sanders** [comment]
This is still broken for my example:

```ini
[file:elsewhere]
source=localhost:/usr/share/dict/words
```

(note on some OSes the words list may be elsewhere)

Looks like the error is surfacing in `RosePopenEvent.__str__`, presumably due to joining strings with bytes read from the file?

> **Note:** the lovely `FIXME: Is this safe? ` comment in this code block relates to "is it safe to read from stdin in this way" - and the answer is no, as I've found out on my branch. Though that can only cause things to hang not fail with str/byte concatination.

**Review by oliver-sanders — APPROVED** [review]
Ok, lets go with this example:

```console
$ echo messy > ~/messy
$ touch flow.cylc
$ cat > rose-suite.conf <<__HERE__
[file:elsewhere]
source=localhost:messy
__HERE__
$ cylc install
```

Tested as working on this branch the same as on Rose2019.

Pytests pass, functional test failures are fun for later ;)

**Review by datamel — APPROVED** [review]
Small style comments.

**Pull request merged**